### PR TITLE
Implement static portable Create methods for ECDsa contract

### PR DESCRIPTION
**Pull request by steveharter**
For #8487 implemented these three methods in ECDsa (s.s.c.Algorithms.dll):
 public static ECDsa Create()
 public static ECDsa Create(ECCurve curve)
 public static ECDsa Create(ECParameters parameters)

Previously they threw a NotImplementedException.

(I expect these to be ported to release branch as well)

Basic work was to move code from cng and openssl assemblies into s.s.c.Algorithms as common code, and have cng and openssl assemblies use that as well. Also hooked up compatible tests to s.s.c.Algorithms.Tests.dll. Followed the same pattern that RSA used to implement its Create method.

The cng side required additional work to abstract away the CngKey for the algorithms.dll only code and for common code.

Also added support to set integer KeySize to a curve after it already was a named curve. Previously setting KeySize after it already had a curve was ignored.

@bartonjs

### CODE REVIEW

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsaCng.Key.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography
+{
+    using Microsoft.Win32.SafeHandles;
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsaCng.Key.cs]: Style: usings go outside the namespace block

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using System.Security.Cryptography.EcDsa.Tests;
+
+namespace System.Security.Cryptography.Cng.Tests
+{
+    public class ECDsaCngImportExportTests : ECDsaTestsBase
+    {
+        [Fact]
+        public static void TestImportKey()
+        {
+            using (CngKey key = CngKey.Import(TestData.Key_ECDiffieHellmanP256, CngKeyBlobFormat.GenericPublicBlob))
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs]: A) Why are we talking about CNG in the platform neutral test library?  B) Why are we talking about ECDH in an ECDSA test?

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                Assert.Equal(CngUIProtectionLevels.None, policy.ProtectionLevel);
+
+                Assert.Equal(null, key.UniqueName);
+            }
+        }
+
+        [Fact]
+        public static void TestImportExportRoundTrip()
+        {
+            using (CngKey key = CngKey.Import(TestData.Key_ECDiffieHellmanP256, CngKeyBlobFormat.GenericPublicBlob))
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs]: Same question

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                byte[] reExported = key.Export(CngKeyBlobFormat.GenericPublicBlob);
+                Assert.Equal<byte>(TestData.Key_ECDiffieHellmanP256, reExported);
+            }
+        }
+
+        [ConditionalTheory(nameof(ECExplicitCurvesSupported)), MemberData(nameof(TestCurves))]
+        public static void TestHashRoundTrip(CurveDef curveDef)
+        {
+            // This test is in the cng only tests because OpenSsl does not provide the hash algorithm
+            using (var cng = new ECDsaCng(curveDef.Curve))
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs]: How are we even talking about this type?

File: src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj [hunk 5]
```diff
@@ -106,5 +131,8 @@
   <ItemGroup>
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Common\Interop\Unix\" />
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/System.Security.Cryptography.Algorithms.Tests.csproj]: Why?

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngTests.cs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Security.Cryptography.EcDsa.Tests;
+using Test.Cryptography;
+using Xunit;
+
+namespace System.Security.Cryptography.Cng.Tests
+{
+    public class ECDsaCngTests : ECDsaTestsBase
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngTests.cs]: This seems to have been copied, but not added to the project.  Probably just needs to be dropped out of the commit?

File: src/System.Security.Cryptography.Algorithms/tests/project.json [hunk 7]
```diff
@@ -2,6 +2,7 @@
   "dependencies": {
     "Microsoft.NETCore.Platforms": "1.0.1-rc3-24117-00",
     "System.IO": "4.1.0-rc3-24117-00",
+    "System.IO.FileSystem": "4.0.1-rc3-24117-00",
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/project.json]: Why do ECDSA tests need FileSystem?

File: src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs [hunk 8]
```diff
@@ -1,68 +1,13 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Internal.Cryptography;
 using Microsoft.Win32.SafeHandles;
-using System.Diagnostics;
-using System.IO;
 
 namespace System.Security.Cryptography
+
```
Comment (bartonjs) [inline src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs]: Please remove the new blank line

**Comment by bartonjs** [comment]
It feels like there's a lot more code duplicated between the CNG and Algorithms implementations than I'd like; but I'd rather have something that works now than something perfect later.

I also feel like I expected more code to show up as "file was renamed with edits" than there was.  I don't know if that's because the files got re-indented with the conditional wrapper type, or github UI is having trouble rendering it.

So, let's get the extra copies of the CNG tests and other small stuff cleaned up, and then we'll be good to go.

It didn't look like any new members were added, is that correct?  (Just double checking)

**Comment by bartonjs** [comment]
Test Outerloop Windows 10 Debug please
Test OuterLoop Windows 7 Debug please

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs (hunk 2 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngImportExportTests.cs]: This test file along with the next accidently got copied to the algorithms location, but not hooked up to the project. Removed

File: src/System.Security.Cryptography.Algorithms/tests/ECDsaCngTests.cs (hunk 6 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/ECDsaCngTests.cs]: Yes

File: src/System.Security.Cryptography.Algorithms/tests/project.json (hunk 7 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/project.json]: PlatformDetection.cs was added to detect the version of Windows, and that uses the File class.

File: src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs (hunk 8 above)
Comment (steveharter) [inline src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs]: Done

**Comment by steveharter** [comment]
No new public members were added; I just ran apigen against algorithms.dll for both unix and windows_nt and didn't detect any. However by doing that I did find that in ref\System.Security.Cryptography.Algorithms.cs the ECParameters struct was missing the StructLayoutAttribute and the 'partial' modifier which I added (shouldn't affect anything; for consistency)

File: src/System.Security.Cryptography.Algorithms/tests/project.json (hunk 7 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/project.json]: Ah, okay, makes sense.

**Comment by steveharter** [comment]
test Innerloop CentOS7.1 Debug Build and Test please

**Comment by bartonjs** [comment]
LGTM with squash.

It looks like you're doctoring this PR up to look like a port-to-rtm PR... that'll have to be a separate PR.  And since we've been having some build stability issues it has been requested that we let this sit in master for a day before porting it to the release branch.

**Pull request merged**